### PR TITLE
Replay hot blocks at startup for protoarray

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
Replays hot blocks in `Store` for `ProtoArray` so that when processing head after a restart, `ProtoArray` does not throw any errors.

Also, switches `processBlock` to use `ReadOnlyStore`, and block state's justified and finalized checkpoints instead of ones from the Store.

### CODE REVIEW

File: protoarray/src/main/java/tech/pegasys/artemis/protoarray/ProtoArrayForkChoiceStrategy.java [hunk 1]
```diff
@@ -113,30 +119,30 @@ public void maybePrune(Bytes32 finalizedRoot) {
...
+          block.getSlot(),
+          blockRoot,
+          store.getBlockRoots().contains(block.getParent_root())
+              ? Optional.of(block.getParent_root())
+              : Optional.empty(),
+          block.getState_root(),
+          store.getBlockState(block.hash_tree_root()).getCurrent_justified_checkpoint().getEpoch(),
+          store.getBlockState(block.hash_tree_root()).getFinalized_checkpoint().getEpoch());
+    }
   }
```
Comment (ajsutton) [inline protoarray/src/main/java/tech/pegasys/artemis/protoarray/ProtoArrayForkChoiceStrategy.java]: You can simplify this with streams to:

```
  private static void processBlocksInStoreAtStartup(ReadOnlyStore store, ProtoArray protoArray) {
    store.getBlockRoots().stream()
        .map(store::getBlock)
        .sorted(Comparator.comparing(BeaconBlock::getSlot))
        .forEach(block -> processBlockAtStartup(store, protoArray, block));
  }

  private static void processBlockAtStartup(
      final ReadOnlyStore store, final ProtoArray protoArray, final BeaconBlock block) {
    Bytes32 blockRoot = block.hash_tree_root();
    protoArray.onBlock(
        block.getSlot(),
        blockRoot,
        store.getBlockRoots().contains(block.getParent_root())
            ? Optional.of(block.getParent_root())
            : Optional.empty(),
        block.getState_root(),
        store.getBlockState(block.hash_tree_root()).getCurrent_justified_checkpoint().getEpoch(),
        store.getBlockState(block.hash_tree_root()).getFinalized_checkpoint().getEpoch());
  }
```

Comment (ajsutton) [reply protoarray/src/main/java/tech/pegasys/artemis/protoarray/ProtoArrayForkChoiceStrategy.java]: You could put the `checkNotNull` back in but I'm not sure it's worth it given that `Store` guarantees that it has a block for every block root (and it literally gets the block roots from the keyset of its map of blocks).

**Review by ajsutton — APPROVED** [review]
LGTM.

File: protoarray/src/main/java/tech/pegasys/artemis/protoarray/ProtoArrayForkChoiceStrategy.java (hunk 1 above)
Comment (cemozerr) [reply protoarray/src/main/java/tech/pegasys/artemis/protoarray/ProtoArrayForkChoiceStrategy.java]: This is neat. Thank you.

**Pull request merged**